src/log-query/src/log_query.rs

```diff
@@ -0,0 +1,323 @@
+                        &naive_date.and_time(NaiveTime::from_hms_opt(0, 0, 0).unwrap()),
+                    );
+                    let end = match *format {
+                        "%Y-%m-%d" => start + Duration::days(1),
+                        "%Y-%m" => {
+                            let next_month = if naive_date.month() == 12 {
+                                NaiveDate::from_ymd_opt(naive_date.year() + 1, 1, 1).unwrap()
+                            } else {
+                                NaiveDate::from_ymd_opt(
+                                    naive_date.year(),
+                                    naive_date.month() + 1,
+                                    1,
+                                )
+                                .unwrap()
+                            };
+                            Utc.from_utc_datetime(&next_month.and_hms_opt(0, 0, 0).unwrap())
+                        }
+                        "%Y" => {
+                            let next_year =
+                                NaiveDate::from_ymd_opt(naive_date.year() + 1, 1, 1).unwrap();
+                            Utc.from_utc_datetime(&next_year.and_hms_opt(0, 0, 0).unwrap())
+                        }
+                        _ => unreachable!(),
+                    };
+                    return Ok((start, Some(end)));
+                }
+            }
+            Err(InvalidDateFormatSnafu {
+                input: s.to_string(),
+            }
+            .build())
+        }
+    }
+
+    /// Util function handles durations like "1 week", "1 month", etc (unimplemented).
+    fn parse_span(s: &str) -> Result<Duration> {
+        // Simplified parsing logic
+        if let Ok(seconds) = s.parse::<i64>() {
+            Ok(Duration::seconds(seconds))
+        } else {
+            Err(InvalidSpanFormatSnafu {
+                input: s.to_string(),
+            }
+            .build())
+        }
+    }
+}
+
+/// Represents a column with filters to query.
+pub struct ColumnFilters {
+    /// Case-sensitive column name to query.
+    pub column_name: String,
+    /// Filters to apply to the column. Can be empty.
+    pub filters: Vec<ContentFilter>,
+}
+
+pub enum ContentFilter {
+    /// Only match the exact content.
+    ///
+    /// For example, if the content is "pale blue dot", the filter "pale" or "pale blue" will match.
+    Exact(String),
+    /// Match the content with a prefix.
+    ///
+    /// For example, if the content is "error message", the filter "err" or "error mess" will match.
+    Prefix(String),
+    /// Match the content with a postfix. Similar to `Prefix`.
+    Postfix(String),
+    /// Match the content with a substring.
+    Contains(String),
+    /// Match the content with a regex pattern. The pattern should be a valid Rust regex.
+    Regex(String),
+    Compound(Vec<ContentFilter>, BinaryOperator),
+}
+
+pub enum BinaryOperator {
+    And,
+    Or,
+}
+
+pub enum Context {
```

Comment:
question: In the context structure, does `usize` refer to the number of lines? If so, can we add the `range` semantics to express the data from a time before and after a specific match? For example, query the logs from 5 minutes before and after a particular keyword appears.